tests/training_success/test_training_success.py

```diff
@@ -0,0 +1,263 @@
+import logging
+from collections import deque
+from pprint import pprint
+from typing import Any, Dict, Tuple
+
+import pandas as pd
+import pytest
+import yaml
+
+from ludwig.api import LudwigModel
+from ludwig.config_sampling.explore_schema import combine_configs, ConfigOption, explore_properties
+from ludwig.config_validation.validation import get_schema
+from ludwig.types import ModelConfigDict
+
+from .configs import (
+    COMBINER_TYPE_TO_COMBINE_FN_MAP,
+    ECD_CONFIG_SECTION_TO_CONFIG,
+    FEATURE_TYPE_TO_CONFIG_FOR_DECODER_LOSS,
+    FEATURE_TYPE_TO_CONFIG_FOR_ENCODER_PREPROCESSING,
+)
+
+
+def defaults_config_generator(
+    feature_type: str, allow_list: str, static_schema: Dict[str, Any] = None
+) -> Tuple[ModelConfigDict, pd.DataFrame]:
+    """Generate combinatorial configs for the defaults section of the Ludwig config.
+
+    Args:
+        feature_type: feature type to explore.
+        allow_list: top-level parameter of the defaults sections that should be included.
+    """
+    assert isinstance(allow_list, str)
+    assert allow_list in {"preprocessing", "encoder", "decoder", "loss"}
+
+    schema = get_schema() if not static_schema else static_schema
+    properties = schema["properties"]["defaults"]["properties"][feature_type]["properties"]
+    raw_entry = deque([ConfigOption(dict(), False)])
+    explored = explore_properties(
+        properties, parent_parameter_path=f"defaults.{feature_type}", dq=raw_entry, allow_list=[allow_list]
+    )
+
+    if allow_list in ["preprocessing", "encoder"]:
+        config = FEATURE_TYPE_TO_CONFIG_FOR_ENCODER_PREPROCESSING[feature_type]
+        config = yaml.safe_load(config)
+    else:  # decoder and loss
+        config = FEATURE_TYPE_TO_CONFIG_FOR_DECODER_LOSS[feature_type]
+        config = yaml.safe_load(config)
+
+    main_config_keys = list(config.keys())
+    for key in main_config_keys:
+        if key not in ["input_features", "output_features"]:
+            del config[key]
```

Comment:
Why is this necessary if all of the configs in `tests/training_success/configs.py` only have `input_features` and `output_features`?